#!/usr/bin/env awk

# *****************************************************************************
	
#	add_PE_info.awk
	
#	This file is part of the ChimPipe pipeline 

#	Copyright (c) 2014 Bernardo Rodríguez-Martín 
#					   Emilio Palumbo 
#					   Sarah djebali 
	
#	Computational Biology of RNA Processing group
#	Department of Bioinformatics and Genomics
#	Centre for Genomic Regulation (CRG)
					   
#	Github repository - https://github.com/Chimera-tools/ChimPipe
	
#	Documentation - https://chimpipe.readthedocs.org/

#	Contact - chimpipe.pipeline@gmail.com
	
#	Licenced under the GNU General Public License 3.0 license.
#******************************************************************************

# Description
###############

# Takes as input 1) File with gene to gene connections from PE information and 2) Chimeric junction matrix generated by Chimsplice. Add to this matrix the information regarding the support of the junctions by PE information. 

#awk -v fileRef:gene_gene_connections.txt -f ~/Awk/add_PE_info.awk distinct_junctions_nbstaggered_nbtotalsplimappings_withmaxbegandend_samechrstr_okgxorder_dist_ss1_ss2_gnlist1_gnlist2_gnname1_gnname2_bt1_bt2_from_split_mappings_part1overA_part2overB_only_A_B_indiffgn_and_inonegn.txt

## input

# file 1 (file ref): pairs_of_diff_gn_supported_by_pereads_nbpereads.txt
# 13 SNORD116 65
# 17 NGS 273
# 19 SNORD116 8

# file 2: chimJunctions_chimSplice.txt
#chr11_61735045_+:chr5_17354684_+	4	8	8	0	61735016	17354719	GT	AG	0	na	na	ENSG00000167996.11	ENSG00000248223.1	FTH1	CTD-2139B15.2	protein_coding	lincRNA	100	100	13	441	SRR201779.4945895_PATHBIO-SOLEXA2_30TUEAAXX:3:77:1778:364_length=53#0/1,SRR201779.1195095_PATHBIO-SOLEXA2_30TUEAAXX:3:18:1340:280_length=53#0/1,SRR201779.137070_PATHBIO-SOLEXA2_30TUEAAXX:3:2:1771:312_length=53#0/1,SRR201779.2821968_PATHBIO-SOLEXA2_30TUEAAXX:3:43:1388:938_length=53#0/2,SRR201779.405498_PATHBIO-SOLEXA2_30TUEAAXX:3:6:1290:23_length=53#0/1,SRR201779.4633065_PATHBIO-SOLEXA2_30TUEAAXX:3:72:418:1024_length=53#0/2,SRR201779.2254852_PATHBIO-SOLEXA2_30TUEAAXX:3:34:1172:781_length=53#0/2,SRR201779.98149_PATHBIO-SOLEXA2_30TUEAAXX:3:2:1130:729_length=53#0/1,

## OUTPUT
# chr11_61735045_+:chr5_17354684_+	4	8	8	0	61735016	17354719	GT	AG	0	na	na	ENSG00000167996.11	ENSG00000248223.1	FTH1	CTD-2139B15.2	protein_coding	100	100	13	441	lincRNA	SRR201779.4945895_PATHBIO-SOLEXA2_30TUEAAXX:3:77:1778:364_length=53#0/1,SRR201779.1195095_PATHBIO-SOLEXA2_30TUEAAXX:3:18:1340:280_length=53#0/1,SRR201779.137070_PATHBIO-SOLEXA2_30TUEAAXX:3:2:1771:312_length=53#0/1,SRR201779.2821968_PATHBIO-SOLEXA2_30TUEAAXX:3:43:1388:938_length=53#0/2,SRR201779.405498_PATHBIO-SOLEXA2_30TUEAAXX:3:6:1290:23_length=53#0/1,SRR201779.4633065_PATHBIO-SOLEXA2_30TUEAAXX:3:72:418:1024_length=53#0/2,SRR201779.2254852_PATHBIO-SOLEXA2_30TUEAAXX:3:34:1172:781_length=53#0/2,SRR201779.98149_PATHBIO-SOLEXA2_30TUEAAXX:3:2:1130:729_length=53#0/1, 2-1:11,


BEGIN{while (getline < fileRef >0)
	{
		PE[$1"-"$2]=$3
	}
}

# Print header
NR==1{
	print $0, "PEsupport";
}

NR>1{
	nbA=split($13,a,",");
	nbB=split($14,b,","); 
	s=""; 
	for (posA=1; posA<=nbA; posA++)
	{
		for (posB=1; posB<=nbB; posB++)
		{
			if (PE[a[posA]"-"b[posB]]!="")
			{
				s=(s)(posA"-"posB)(":")(PE[a[posA]"-"b[posB]])(",");
			}
			else
			{
				if (PE[b[posB]"-"a[posA]]!="")
				{
					s=(s)(posA"-"posB)(":")(PE[b[posB]"-"a[posA]])(",");
				}
			}
		}
	}
	if (s=="")
	{	
		s=".";
	} 
	print $0, s;
} 
